src/index.js

```diff
@@ -0,0 +1,220 @@
+      exec(`${circomBinary} ${filepath} -o ${output}`, error => {
+        if (error) {
+          return reject(error);
+        }
+
+        resolve();
+      });
+    });
+  }
+
+  async generateProofs() {
+    this.embark.logger.info('Generating proofs...');
+
+    await Promise.all(
+      (await fs.readdir(this.compiledCircuitsPath))
+        .filter(filename => path.extname(filename) === Extension.Json)
+        .map(filename => this.generateProof(filename))
+    );
+  }
+
+  async generateProof(filename) {
+    const filepath = path.join(this.compiledCircuitsPath, filename);
+    const basename = path.basename(filename, Extension.Json);
+
+    const circuit = await this.getCircuit(filepath);
+    const setup = await this.generateSetup(circuit, basename);
+
+    const input = this.circuitsConfig.inputs[basename];
+    if (!input) {
+      return;
+    }
+
+    const witness = circuit.calculateWitness(input);
+    const { proof, publicSignals } = zkSnark.original.genProof(
+      setup.vk_proof,
+      witness
+    );
+    if (zkSnark.original.isValid(setup.vk_verifier, proof, publicSignals)) {
+      return this.generateVerifier(basename);
+    }
+
+    throw new Error(
+      `The proof is not valid for ${basename} with inputs: ${JSON.stringify(
+        input
+      )}`
+    );
+  }
+
+  async getCircuit(filepath) {
+    const definition = JSON.parse(await fs.readFile(filepath, 'utf8'));
+    return new zkSnark.Circuit(definition);
+  }
+
+  async generateSetup(circuit, basename) {
+    const setup = zkSnark.original.setup(circuit);
+    await Promise.all([
+      fs.writeFile(
+        this.proofFilepath(basename),
+        JSON.stringify(setup.vk_proof),
+        'utf8'
+      ),
+      fs.writeFile(
+        this.verifierFilepath(basename),
+        JSON.stringify(setup.vk_verifier),
+        'utf8'
+      )
+    ]);
+
+    return setup;
+  }
+
+  generateVerifier(basename) {
+    return new Promise((resolve, reject) => {
+      const source = this.verifierFilepath(basename);
+      const output = this.verifierContractPath(basename);
+      exec(
+        `${snarkjsBinary} generateverifier --vk ${source} -v ${output}`,
+        error => {
+          if (error) {
+            return reject(error);
+          }
+
+          resolve();
+        }
+      );
+    });
+  }
+
+  async addVerifiersToContracts() {
+    (await fs.readdir(this.compiledCircuitsPath))
+      .filter(filename => path.extname(filename) === Extension.Solidity)
+      .map(filename => this.addVerifierToContracts(filename));
+  }
+
+  addVerifierToContracts(filename) {
+    // won't work as desired with "dev embark" (built in monorepo) until after
+    // first prerelease of embark v5
+    if (semver(this.embark.version).major >= 5 /* || true */) {
+      // get rid of `|| true`
+      this.embark.addContractFile(
+        path.join(this.compiledCircuitsPath, filename)
+      );
+      // this.embark.logger.warn(require('util').inspect(this.embark.contractsFiles));
```

Comment:
Couple of comments.
Also, I think `addContractFile` existed in Embark 4 too. Might just be that we didn't use it?

Reply:
That's something I wasn't sure about. And neither approach is working with the `master` branch currently. I'm hoping @iurimatias and/or someone else can help me figure it out.

Reply:
`addContractFile` is [definitely in `master` prior to the v5 refactor](https://github.com/embark-framework/embark/blob/v4.1.1/packages/embark/src/lib/core/plugin.js#L172).

So I just tried with a production install of embark@4.1.1 to use `this.embark.addContractFile` instead of `this.embark.events.request('config:contractsFiles:add', ...)` but it doesn't work. I added a logger line that runs after `addContractFile` is called:

```
this.embark.logger.warn(require('util').inspect(this.embark.contractsFiles));
```

And in the cli dashboard I see:

```
embark-snark: [ '/Users/michael/dapps/4.1.1/snarky/.embark/snarks/multiplier.sol' ]
```

So it does push the path onto the array, but there's no effect. `embark run` eventually reports `no contracts found`.

I tried putting `this.embark.addPluginType('contractFiles');` in the plugin's constructor thinking maybe it was happening too late, but that didn't have any effect.

My guess is that `addContractFile` is bugged, but I'm not sure how long that's been the case. Maybe we can fix it and make a patch release for 4.x in addition to having it working in v5.

Reply:
Yeah, does sound like a bug. Could be worth a patch if we use it here

Reply:
For the record, I didn't end up fixing `addContractFile`; based on Iuri's input I added `:before` and `:after` hooks for a compiler event in https://github.com/embark-framework/embark/pull/1878.

The reason for that is to play nicely with the control-flow in this function:

https://github.com/embark-framework/embark/blob/master/packages/embark/src/cmd/cmd_controller.js#L915-L924

Adding files to the array that's modified via `addContractFile` or requesting `config:contractsFiles:add` can't have any effect after L916. So, short of refactoring that arrangement, hooks on `compiler:contracts:compile` were the way to go.